compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1567,6 +1573,92 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    hir::Node::Pat(hir::Pat { hir_id: pat_hir_id, kind: hir::PatKind::Tuple(pats, ..), .. }) => {
+                        let Some(hir::Node::Local(hir::Local { init: Some(init), .. })) =
+                                self.tcx.hir().find(self.tcx.hir().parent_id(*pat_hir_id)) else {
+                            return expr;
+                        };
+
+                        let find_init = |init_tup: &'b [Expr<'b>]| {
+                            pats.iter().enumerate().filter(|x| x.1.hir_id == *hir_id).map(|(i,_)| init_tup.get(i).unwrap()).next()
+                        };
+
+                        match init.kind {
+                            ExprKind::Tup(init_tup) => {
+                                if let Some(init) = find_init(init_tup) {
+                                    self.note_type_is_not_clone_inner_expr(init)
+                                } else {
+                                    expr
+                                }
+                            }
+                            ExprKind::Block(_, _) => {
+                                let block_expr = init.peel_blocks();
+                                if let Expr { kind: ExprKind::Tup(init_tup), .. } = block_expr
+                                    && let Some(init) = find_init(init_tup)
+                                {
+                                    self.note_type_is_not_clone_inner_expr(init)
+                                } else {
+                                    expr
+                                }
+                            }
+                            _ => expr
+                        }
+                    }
+                    _ => expr
+                }
+            }
+            // If we're a closure recurse back into that closure
+            hir::ExprKind::Closure(hir::Closure { body: body_id, .. }) => {
+                let hir::Body { value: body_expr, .. } = self.tcx.hir().body(*body_id);
+                self.note_type_is_not_clone_inner_expr(body_expr)
+            }
+            // Similarly recurse into a block
+            hir::ExprKind::Block(hir::Block { expr: Some(block_expr), .. }, _) => {
+                self.note_type_is_not_clone_inner_expr(block_expr)
+            }
+            hir::ExprKind::Call(Expr { kind: call_expr_kind, .. }, _) => {
+                if let hir::ExprKind::Path(hir::QPath::Resolved( None, call_expr_path,)) = call_expr_kind
+                    && let hir::Path { segments: [_], res: crate::Res::Local(binding), .. } = call_expr_path
+                    && let Some(hir::Node::Pat(hir::Pat { hir_id, .. })) = self.tcx.hir().find(*binding)
+                    && let Some(closure) = self.tcx.hir().find(self.tcx.hir().parent_id(*hir_id))
+                    && let hir::Node::Local(hir::Local { init: Some(init), .. }) = closure
+                {
+                    self.note_type_is_not_clone_inner_expr(init)
```

Comment:
this seems to be specific to the `clone_thing9` test case? If this is so and this is either a closure or something else that we don't care about, I'd prefer this to extract the closure body here instead of in a match case above. Therefore you should remove that match arm and move that here.

Also, could you add some comments on this `if let` chain, to explain what we are trying to do here (lookup the source of a closure call)?